source/language/scope.rst

```diff
@@ -0,0 +1,318 @@
+      //   - 'c': type 'const int', value 3
+      //   - 'd': type 'const int', value 4
+      //   - 'q': type 'qubit[4]', a virtual, run-time-known qubit register.
+      //   - 'alias': alias for the first three qubits of 'q'.
+      //   - 'new_variable': type 'const float[64]', value 1.5
+      //   - the other built-in identifiers like 'U' and 'pi'
+      //   - the available hardware qubits like '$0'.
+   }
+
+
+.. _scope-block:
+
+Block scope
+-----------
+
+Certain control-flow operations introduce their own local scope.  These
+operations are:
+
+* ``for`` loops,
+* ``while`` loops,
+* ``if`` and ``else`` blocks,
+* ``box`` statements.
+
+These scopes inherit *all* variables that are in scope in the immediately
+containing scope.  Unlike subroutines and gate scopes, this includes variables
+that are not ``const``.  This is broadly similar how these constructs behave in
+other procedural languages, such as C.
+
+The iteration variable of a ``for`` loop has lifetime and visibility as if it
+were declared as the first statement in the body of that loop.  It is not
+accessible after the body of the loop.
+
+The blocks associated with ``if`` and its corresponding ``else`` define two
+different scopes; the variables and definitions are not shared between them.
+
+As with subroutine scopes, variables defined locally in these scopes (including
+the for-loop iteration variable) may shadow variables with the same name in
+outer scopes.  When the defining scope of a shadowing variable ends, the
+previous variable (which was shadowed) becomes accessible again.  Qubits and
+arrays cannot be declared within local block scopes, but aliases can.
+
+Some further examples:
+
+.. code-block:: c
+   :linenos:
+
+   OPENQASM 3.0
+
+   int ii = 100;        // 'ii' is declared in the global scope.
+   qubit[5] q;          // 'q' is declared in the global scope.
+   let alias = q[0:2];  // alias 'alias' is declared in the global scope.
+
+   if (true) {
+     ii *= 2;    // This is the global 'ii', which now has the value 200.
+
+     // A local variable 'ii' is declared, which shadows the global definition.
+     // The global 'ii' is no longer accessible until this scope ends.
+     int ii = 1;
+     // The local variable 'ii' is modified, and now has the value 2.
+     ii *= 2;
+   }
+
+   // The local 'ii' went out of scope at the conclusion of the 'if' block, and
+   // the previous 'ii' defined on line 3 is accessible again.
+   ii *= 2;  // global 'ii' is now 400.
+
+   uint sum = 0;
+   for uint ii in [1:10] {
+     // The global 'ii' is shadowed by the iteration variable 'ii', which also
+     // has a different type.  The outer 'sum' is still accessible.
+
+     sum += ii;  // Iteration variable 'ii' is added to global 'sum'
+     if (sum > 10) {
+       float ii = 10.0; // For-loop iteration variable shadowed.
+       sum += uint(ii*2);
+     } else {
+       sum += ii;      // 'ii' is the for-loop iteration variable.
+     }
+
```

Comment:
**thought**: Perhaps we should add an example like this 
```c
    int ii = 5;
    for uint i in [1:10] {
        ii += 2;
        int ii = 2;
        ...
    }
```

Here is equivalent C output 
```
iteration i = 0
ii before declaration = 7
ii after declaration = 2

iteration i = 1
ii before declaration = 9
ii after declaration = 2
```

Reply:
Looks good.